test/blackbox/common/BlackboxTestsLivelinessQos.cpp

```diff
@@ -1946,6 +1948,54 @@ TEST(LivelinessTests, Detect_Deadlock_ManualByParticipant_Intraprocess)
     // Test failure is due to timeout
 }
 
+// Regression test of Refs #20584, github issue #4373
+TEST(LivelinessTests, Reader_Successfully_Asserts_Liveliness_on_a_Disconnected_Writer)
+{
+    // Create a TestTransport to simulate a network shutdown (Ctrl+C)
+    auto testTransport = std::make_shared<eprosima::fastdds::rtps::test_UDPv4TransportDescriptor>();
+
+    // Create two writer participants
+    PubSubWriter<HelloWorldPubSubType> writer_1(TEST_TOPIC_NAME);
+    PubSubWriter<HelloWorldPubSubType> writer_2(TEST_TOPIC_NAME + "2");
+
+    // Create a reader participant containing 2 subscribers and readers
+    PubSubParticipant<HelloWorldPubSubType> reader(0, 2, 0, 2);
+
+    reader.init_participant();
+    // Define the reader's lease duration in 1.6 secs
+    reader.sub_liveliness_lease_duration(eprosima::fastrtps::Time_t(1, 600000000));
+
+    // Create Subscribers and readers, one for each writer
+    reader.sub_topic_name(TEST_TOPIC_NAME);
+    reader.init_subscriber(0);
+    reader.sub_topic_name(TEST_TOPIC_NAME + "2");
+    reader.init_subscriber(1);
+
+    // Create writers
+    writer_1.disable_builtin_transport()
+            .add_user_transport_to_pparams(testTransport)
+            .liveliness_lease_duration(eprosima::fastrtps::Time_t(1, 0))
+            .liveliness_kind(eprosima::fastdds::dds::AUTOMATIC_LIVELINESS_QOS)
+            .liveliness_announcement_period(eprosima::fastrtps::Time_t(0, 900000000))
+            .init();
+
+    writer_2.liveliness_lease_duration(eprosima::fastrtps::Time_t(1, 0))
```

Comment:
Please ensure that the participant lease duration is greater than the liveliness lease duration. It could even be INFINITE to ensure the reader liveliness is the one being checked.